diff --git a/lib/init.lua b/lib/init.lua
index 63bb787..a134de1 100644
--- a/lib/init.lua
+++ b/lib/init.lua
@@ -1,23 +1,26 @@
+--!strict
 -- t: a runtime typechecker for Roblox
 
 local t = {}
 
-function t.type(typeName)
+type check = (any) -> (boolean, string?)
+
+function t.type(typeName: string): check
 	return function(value)
 		local valueType = type(value)
 		if valueType == typeName then
-			return true
+			return true, nil
 		else
 			return false, string.format("%s expected, got %s", typeName, valueType)
 		end
 	end
 end
 
-function t.typeof(typeName)
+function t.typeof(typeName: string): check
 	return function(value)
 		local valueType = typeof(value)
 		if valueType == typeName then
-			return true
+			return true, nil
 		else
 			return false, string.format("%s expected, got %s", typeName, valueType)
 		end
@@ -31,9 +34,9 @@ end
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.any(value)
+function t.any(value: any): (boolean, string?)
 	if value ~= nil then
-		return true
+		return true, nil
 	else
 		return false, "any expected, got nil"
 	end
@@ -131,11 +134,11 @@ t.vector = t.type("vector")
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.number(value)
+function t.number(value: any): (boolean, string?)
 	local valueType = typeof(value)
 	if valueType == "number" then
 		if value == value then
-			return true
+			return true, nil
 		else
 			return false, "unexpected NaN value"
 		end
@@ -151,11 +154,11 @@ end
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.nan(value)
+function t.nan(value: any): (boolean, string?)
 	local valueType = typeof(value)
 	if valueType == "number" then
 		if value ~= value then
-			return true
+			return true, nil
 		else
 			return false, "unexpected non-NaN value"
 		end
@@ -515,7 +518,7 @@ t.Vector3int16 = t.typeof("Vector3int16")
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.literal(...)
+function t.literal(...): check
 	local size = select("#", ...)
 	if size == 1 then
 		local literal = ...
@@ -524,7 +527,7 @@ function t.literal(...)
 				return false, string.format("expected %s, got %s", tostring(literal), tostring(value))
 			end
 
-			return true
+			return true, nil
 		end
 	else
 		local literals = {}
@@ -550,10 +553,10 @@ t.exactly = t.literal
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.keyOf(keyTable)
+function t.keyOf(keyTable: { [any]: any }): check
 	local keys = {}
 	local length = 0
-	for key in pairs(keyTable) do
+	for key in keyTable do
 		length = length + 1
 		keys[length] = key
 	end
@@ -568,10 +571,10 @@ end
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.valueOf(valueTable)
+function t.valueOf(valueTable: { [any]: any }): check
 	local values = {}
 	local length = 0
-	for _, value in pairs(valueTable) do
+	for _, value in valueTable do
 		length = length + 1
 		values[length] = value
 	end
@@ -586,14 +589,14 @@ end
 
 	@returns True iff the condition is satisfied, false otherwise
 **--]]
-function t.integer(value)
+function t.integer(value: any): (boolean, string?)
 	local success, errMsg = t.number(value)
 	if not success then
 		return false, errMsg or ""
 	end
 
 	if value % 1 == 0 then
-		return true
+		return true, nil
 	else
 		return false, string.format("integer expected, got %s", value)
 	end
@@ -606,7 +609,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberMin(min)
+function t.numberMin(min: number): check
 	return function(value)
 		local success, errMsg = t.number(value)
 		if not success then
@@ -614,9 +617,9 @@ function t.numberMin(min)
 		end
 
 		if value >= min then
-			return true
+			return true, nil
 		else
-			return false, string.format("number >= %s expected, got %s", min, value)
+			return false, string.format("number >= %f expected, got %s", min, value)
 		end
 	end
 end
@@ -628,7 +631,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberMax(max)
+function t.numberMax(max: number): check
 	return function(value)
 		local success, errMsg = t.number(value)
 		if not success then
@@ -636,9 +639,9 @@ function t.numberMax(max)
 		end
 
 		if value <= max then
-			return true
+			return true, nil
 		else
-			return false, string.format("number <= %s expected, got %s", max, value)
+			return false, string.format("number <= %f expected, got %s", max, value)
 		end
 	end
 end
@@ -650,7 +653,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberMinExclusive(min)
+function t.numberMinExclusive(min: number): check
 	return function(value)
 		local success, errMsg = t.number(value)
 		if not success then
@@ -658,9 +661,9 @@ function t.numberMinExclusive(min)
 		end
 
 		if min < value then
-			return true
+			return true, nil
 		else
-			return false, string.format("number > %s expected, got %s", min, value)
+			return false, string.format("number > %f expected, got %s", min, value)
 		end
 	end
 end
@@ -672,7 +675,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberMaxExclusive(max)
+function t.numberMaxExclusive(max: number): check
 	return function(value)
 		local success, errMsg = t.number(value)
 		if not success then
@@ -680,9 +683,9 @@ function t.numberMaxExclusive(max)
 		end
 
 		if value < max then
-			return true
+			return true, nil
 		else
-			return false, string.format("number < %s expected, got %s", max, value)
+			return false, string.format("number < %f expected, got %s", max, value)
 		end
 	end
 end
@@ -709,7 +712,7 @@ t.numberNegative = t.numberMaxExclusive(0)
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberConstrained(min, max)
+function t.numberConstrained(min: number, max: number): check
 	assert(t.number(min))
 	assert(t.number(max))
 	local minCheck = t.numberMin(min)
@@ -726,7 +729,7 @@ function t.numberConstrained(min, max)
 			return false, maxErrMsg or ""
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -738,7 +741,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.numberConstrainedExclusive(min, max)
+function t.numberConstrainedExclusive(min: number, max: number): check
 	assert(t.number(min))
 	assert(t.number(max))
 	local minCheck = t.numberMinExclusive(min)
@@ -755,7 +758,7 @@ function t.numberConstrainedExclusive(min, max)
 			return false, maxErrMsg or ""
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -766,7 +769,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.match(pattern)
+function t.match(pattern: string): check
 	assert(t.string(pattern))
 	return function(value)
 		local stringSuccess, stringErrMsg = t.string(value)
@@ -774,11 +777,11 @@ function t.match(pattern)
 			return false, stringErrMsg
 		end
 
-		if string.match(value, pattern) == nil then
+		if not string.match(value, pattern) then
 			return false, string.format("%q failed to match pattern %q", value, pattern)
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -789,16 +792,16 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.optional(check)
+function t.optional(check: check): check
 	assert(t.callback(check))
 	return function(value)
 		if value == nil then
-			return true
+			return true, nil
 		end
 
 		local success, errMsg = check(value)
 		if success then
-			return true
+			return true, nil
 		else
 			return false, string.format("(optional) %s", errMsg or "")
 		end
@@ -812,18 +815,18 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.tuple(...)
+function t.tuple(...: check): check
 	local checks = { ... }
 	return function(...)
 		local args = { ... }
-		for i, check in ipairs(checks) do
+		for i, check in checks do
 			local success, errMsg = check(args[i])
 			if success == false then
-				return false, string.format("Bad tuple index #%s:\n\t%s", i, errMsg or "")
+				return false, string.format("Bad tuple index #%d:\n\t%s", i, errMsg or "")
 			end
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -834,7 +837,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.keys(check)
+function t.keys(check: check): check
 	assert(t.callback(check))
 	return function(value)
 		local tableSuccess, tableErrMsg = t.table(value)
@@ -842,14 +845,14 @@ function t.keys(check)
 			return false, tableErrMsg or ""
 		end
 
-		for key in pairs(value) do
+		for key in value do
 			local success, errMsg = check(key)
 			if success == false then
 				return false, string.format("bad key %s:\n\t%s", tostring(key), errMsg or "")
 			end
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -860,7 +863,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.values(check)
+function t.values(check: check): check
 	assert(t.callback(check))
 	return function(value)
 		local tableSuccess, tableErrMsg = t.table(value)
@@ -868,14 +871,14 @@ function t.values(check)
 			return false, tableErrMsg or ""
 		end
 
-		for key, val in pairs(value) do
+		for key, val in value do
 			local success, errMsg = check(val)
 			if success == false then
 				return false, string.format("bad value for key %s:\n\t%s", tostring(key), errMsg or "")
 			end
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -887,7 +890,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.map(keyCheck, valueCheck)
+function t.map(keyCheck: check, valueCheck: check): check
 	assert(t.callback(keyCheck))
 	assert(t.callback(valueCheck))
 	local keyChecker = t.keys(keyCheck)
@@ -904,7 +907,7 @@ function t.map(keyCheck, valueCheck)
 			return false, valueErr or ""
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -915,20 +918,20 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.set(valueCheck)
+function t.set(valueCheck: check): check
 	return t.map(valueCheck, t.literal(true))
 end
 
 do
 	local arrayKeysCheck = t.keys(t.integer)
---[[**
+	--[[**
 		ensures value is an array and all values of the array match check
 
 		@param check The check to compare all values with
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.array(check)
+	function t.array(check: check): check
 		assert(t.callback(check))
 		local valuesCheck = t.values(check)
 
@@ -939,14 +942,14 @@ do
 			end
 
 			-- # is unreliable for sparse arrays
-			-- Count upwards using ipairs to avoid false positives from the behavior of #
+			-- Count upwards using `for _ in` to avoid false positives from the behavior of #
 			local arraySize = 0
 
-			for _ in ipairs(value) do
+			for _ in value do
 				arraySize = arraySize + 1
 			end
 
-			for key in pairs(value) do
+			for key in value do
 				if key < 1 or key > arraySize then
 					return false, string.format("[array] key %s must be sequential", tostring(key))
 				end
@@ -957,18 +960,18 @@ do
 				return false, string.format("[array] %s", valueErrMsg or "")
 			end
 
-			return true
+			return true, nil
 		end
 	end
 
---[[**
+	--[[**
 		ensures value is an array of a strict makeup and size
 
 		@param check The check to compare all values with
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.strictArray(...)
+	function t.strictArray(...): check
 		local valueTypes = { ... }
 		assert(t.array(t.callback)(valueTypes))
 
@@ -983,35 +986,35 @@ do
 				return false, string.format("[strictArray] Array size exceeds limit of %d", #valueTypes)
 			end
 
-			for idx, typeFn in pairs(valueTypes) do
+			for idx, typeFn in valueTypes do
 				local typeSuccess, typeErrMsg = typeFn(value[idx])
 				if not typeSuccess then
 					return false, string.format("[strictArray] Array index #%d - %s", idx, typeErrMsg)
 				end
 			end
 
-			return true
+			return true, nil
 		end
 	end
 end
 
 do
 	local callbackArray = t.array(t.callback)
---[[**
+	--[[**
 		creates a union type
 
 		@param ... The checks to union
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.union(...)
+	function t.union(...): check
 		local checks = { ... }
 		assert(callbackArray(checks))
 
 		return function(value)
-			for _, check in ipairs(checks) do
+			for _, check in checks do
 				if check(value) then
-					return true
+					return true, nil
 				end
 			end
 
@@ -1019,35 +1022,35 @@ do
 		end
 	end
 
---[[**
+	--[[**
 		Alias for t.union
 	**--]]
 	t.some = t.union
 
---[[**
+	--[[**
 		creates an intersection type
 
 		@param ... The checks to intersect
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.intersection(...)
+	function t.intersection(...: check): check
 		local checks = { ... }
 		assert(callbackArray(checks))
 
 		return function(value)
-			for _, check in ipairs(checks) do
+			for _, check in checks do
 				local success, errMsg = check(value)
 				if not success then
 					return false, errMsg or ""
 				end
 			end
 
-			return true
+			return true, nil
 		end
 	end
 
---[[**
+	--[[**
 		Alias for t.intersection
 	**--]]
 	t.every = t.intersection
@@ -1055,14 +1058,14 @@ end
 
 do
 	local checkInterface = t.map(t.any, t.callback)
---[[**
+	--[[**
 		ensures value matches given interface definition
 
 		@param checkTable The interface definition
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.interface(checkTable)
+	function t.interface(checkTable: { [any]: check }): check
 		assert(checkInterface(checkTable))
 		return function(value)
 			local tableSuccess, tableErrMsg = t.table(value)
@@ -1070,25 +1073,25 @@ do
 				return false, tableErrMsg or ""
 			end
 
-			for key, check in pairs(checkTable) do
+			for key, check in checkTable do
 				local success, errMsg = check(value[key])
 				if success == false then
 					return false, string.format("[interface] bad value for %s:\n\t%s", tostring(key), errMsg or "")
 				end
 			end
 
-			return true
+			return true, nil
 		end
 	end
 
---[[**
+	--[[**
 		ensures value matches given interface definition strictly
 
 		@param checkTable The interface definition
 
 		@returns A function that will return true iff the condition is passed
 	**--]]
-	function t.strictInterface(checkTable)
+	function t.strictInterface(checkTable: { [any]: check }): check
 		assert(checkInterface(checkTable))
 		return function(value)
 			local tableSuccess, tableErrMsg = t.table(value)
@@ -1096,20 +1099,20 @@ do
 				return false, tableErrMsg or ""
 			end
 
-			for key, check in pairs(checkTable) do
+			for key, check in checkTable do
 				local success, errMsg = check(value[key])
 				if success == false then
 					return false, string.format("[interface] bad value for %s:\n\t%s", tostring(key), errMsg or "")
 				end
 			end
 
-			for key in pairs(value) do
+			for key in value do
 				if not checkTable[key] then
 					return false, string.format("[interface] unexpected field %q", tostring(key))
 				end
 			end
 
-			return true
+			return true, nil
 		end
 	end
 end
@@ -1121,7 +1124,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.instanceOf(className, childTable)
+function t.instanceOf(className: string, childTable: { [string]: check }?): check
 	assert(t.string(className))
 
 	local childrenCheck
@@ -1134,6 +1137,7 @@ function t.instanceOf(className, childTable)
 		if not instanceSuccess then
 			return false, instanceErrMsg or ""
 		end
+		assert(typeof(value) == "Instance", "t.Instance returned true for non-Instance value")
 
 		if value.ClassName ~= className then
 			return false, string.format("%s expected, got %s", className, value.ClassName)
@@ -1146,7 +1150,7 @@ function t.instanceOf(className, childTable)
 			end
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -1159,7 +1163,7 @@ t.instance = t.instanceOf
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.instanceIsA(className, childTable)
+function t.instanceIsA(className: string, childTable: { [string]: check }?): check
 	assert(t.string(className))
 
 	local childrenCheck
@@ -1172,6 +1176,7 @@ function t.instanceIsA(className, childTable)
 		if not instanceSuccess then
 			return false, instanceErrMsg or ""
 		end
+		assert(typeof(value) == "Instance", "t.Instance returned true for non-Instance value")
 
 		if not value:IsA(className) then
 			return false, string.format("%s expected, got %s", className, value.ClassName)
@@ -1184,7 +1189,7 @@ function t.instanceIsA(className, childTable)
 			end
 		end
 
-		return true
+		return true, nil
 	end
 end
 
@@ -1195,7 +1200,7 @@ end
 
 	@returns A function that will return true iff the condition is passed
 **--]]
-function t.enum(enum)
+function t.enum(enum: any): check
 	assert(t.Enum(enum))
 	return function(value)
 		local enumItemSuccess, enumItemErrMsg = t.EnumItem(value)
@@ -1204,7 +1209,7 @@ function t.enum(enum)
 		end
 
 		if value.EnumType == enum then
-			return true
+			return true, nil
 		else
 			return false, string.format("enum of %s expected, got enum of %s", tostring(enum), tostring(value.EnumType))
 		end
@@ -1214,7 +1219,7 @@ end
 do
 	local checkWrap = t.tuple(t.callback, t.callback)
 
---[[**
+	--[[**
 		wraps a callback in an assert with checkArgs
 
 		@param callback The function to wrap
@@ -1238,7 +1243,7 @@ end
 
 	@returns A function that simply wraps the given check in an assert
 **--]]
-function t.strict(check)
+function t.strict(check: check)
 	return function(...)
 		assert(check(...))
 	end
@@ -1247,7 +1252,7 @@ end
 do
 	local checkChildren = t.map(t.string, t.callback)
 
---[[**
+	--[[**
 		Takes a table where keys are child names and values are functions to check the children against.
 		Pass an instance tree into the function.
 		If at least one child passes each check, the overall check passes.
@@ -1258,7 +1263,7 @@ do
 
 		@returns A function that checks an instance tree
 	**--]]
-	function t.children(checkTable)
+	function t.children(checkTable: { [string]: check }): check
 		assert(checkChildren(checkTable))
 
 		return function(value)
@@ -1266,9 +1271,10 @@ do
 			if not instanceSuccess then
 				return false, instanceErrMsg or ""
 			end
+			assert(typeof(value) == "Instance", "t.Instance returned true for non-Instance value")
 
 			local childrenByName = {}
-			for _, child in ipairs(value:GetChildren()) do
+			for _, child in value:GetChildren() do
 				local name = child.Name
 				if checkTable[name] then
 					if childrenByName[name] then
@@ -1279,16 +1285,16 @@ do
 				end
 			end
 
-			for name, check in pairs(checkTable) do
+			for name, check in checkTable do
 				local success, errMsg = check(childrenByName[name])
 				if not success then
 					return false, string.format("[%s.%s] %s", value:GetFullName(), name, errMsg or "")
 				end
 			end
 
-			return true
+			return true, nil
 		end
 	end
 end
 
-return t
+return t
\ No newline at end of file
